Base/src/main/java/io/deephaven/base/ringbuffer/CharRingBuffer.java

```diff
@@ -15,25 +18,32 @@
 public class CharRingBuffer implements Serializable {
     protected final boolean growable;
     protected char[] storage;
-    protected int head, tail;
+    protected int head, tail, size;
 
     private void grow(int increase) {
         if (growable) {
-            final int size = size();
+            // assert that we are not asking for the impossible
+            Assert.eqTrue(ArrayUtil.MAX_ARRAY_SIZE - increase >= size, "CharRingBuffer size <= MAX_ARRAY_SIZE");
 
-            final int minLength = size + increase + 1;
+            final int minLength = size + increase;
             int newLength = storage.length * 2;
             while (newLength < minLength) {
-                newLength = newLength * 2;
+                newLength = Math.min(newLength * 2, ArrayUtil.MAX_ARRAY_SIZE);
             }
             char[] newStorage = new char[newLength];
-            if (tail >= head) {
-                System.arraycopy(storage, head, newStorage, 0, size);
-            } else {
-                final int firstCopyLen = storage.length - head;
-                System.arraycopy(storage, head, newStorage, 0, firstCopyLen);
-                System.arraycopy(storage, 0, newStorage, firstCopyLen, size - firstCopyLen);
-            }
+
+            // three scenarios: size is zero so nothing to copy, head is before tail so only one copy needed, head
+            // after tail so two copies needed.  Assuming that copying zero bytes is a fast operation, we will always
```

Comment:
I'm not sure this is a valid assumption. Our chunk copying code has a threshold, even. Maybe ask Corey to dredge up his memories of any experimentation.